File: Quicksort.scala
import java.lang.Math

object Quicksort
{
  def quicksort(xs: Array[Int])
  {
    def swap(i: Int, j: Int)
    {
      val t = xs(i);
      xs(i) = xs(j);
      xs(j) = t
    }

    def sort1(left: Int, right: Int)
    {
      val pivot = xs((left + right) >> 1)
      var i = left;
      var j = right
      while (i <= j)
      {
        while (xs(i) < pivot) i += 1
        while (xs(j) > pivot) j -= 1
        if (i <= j)
        {
          swap(i, j)
          i += 1
          j -= 1
        }
      }
      if (left < j)
      {
        sort1(left, j)
      }
      if (right > i)
      {
        sort1(i, right)
      }
    }
    sort1(0, xs.length - 1)
  }

  def doSort(len: Int):Long = 
  {
    var i: Int = 0
    val a: Array[Int] = new Array[Int](len)

    for (e <- a)
    {
      a(i) = (Math.random() * len).asInstanceOf[Int];
      if (i % 2 == 0)
      {
        a(i) = -a(i)
      }
      i = i + 1
    }

    val start = System.currentTimeMillis()
    quicksort(a)
    val diff = System.currentTimeMillis() - start

    for (i <- 1 to a.length - 1)
    {
      if (a(i-1) > a(i))
      {
        System.err.println("did not sort")
        System.exit(1)
      }
    }

    return diff
  }

  def main(args: Array[String])
  {
    if (args.length != 1)
    {
      System.err.println("usage: scala Quicksort.scala ARRAYLENGTH")
      System.exit(1)
    }

    val len = Integer.parseInt(args(0))

    var min = Long.MaxValue
    var time : Long = 0

    for (i <- 0 to 5)
    {
      time = doSort(len)
      if (time < min)
      {
        min = time
      }
    }

    println("Scala: " + min + " ms (" + len + ")")
  }
}

File: Quicksort.java
public class Quicksort
{
   public static void swap(int[] arr, int i, int j)
   {
      int temp = arr[i];
      arr[i] = arr[j];
      arr[j] = temp;
   }

   public static void quicksort(int[] arr, int left, int right)
   {
      int pivot = arr[(left + right) >> 1];
      int i = left;
      int j = right;
      while (i <= j)
      {
         while (arr[i] < pivot)
            i++;
         while (arr[j] > pivot)
            j--;
         if (i <= j)
         {
            swap(arr, i++, j--);
         }
      }
      if (left < j)
      {
         quicksort(arr, left, j);
      }
      if (i < right)
      {
         quicksort(arr, i, right);
      }
   }

   public static long sort(int len)
   {
      int[] arr = new int[len];
      for (int i = 0; i < arr.length; i++)
      {
         arr[i] = (int)(Math.random() * len);
         if (i % 2 == 0)
         {
            arr[i] = -arr[i];
         }
      }

      long start = System.currentTimeMillis();
      quicksort(arr, 0, arr.length - 1);
      long diff = System.currentTimeMillis() - start;

      for (int i = 1 ; i < arr.length ; i++)
      {
         if (arr[i-1] > arr[i])
         {
            throw new AssertionError("not sorted!!");
         }
      }

      return diff;
   }

   public static void main(String[] args)
   {
      if (args.length != 1)
      {
         System.err.println("usage: java Quicksort ARRAYLENGTH");
         System.exit(1);
      }

      int len = Integer.parseInt(args[0]);

      long time = Long.MAX_VALUE;
      
      for (int i = 0 ; i < 5 ; i++)
      {
         time = Math.min(time, sort(len));
      }

      System.out.println("java " + Quicksort.class.getName() + ": " + time + " ms (" + len + ")");
   }
}